components/Palettes/src/MDCPalettes.h

```diff
@@ -215,4 +215,24 @@ CG_EXTERN const MDCPaletteAccent _Nonnull MDCPaletteAccent700Name;
 /** The A700 accent color, the darkest accent color. */
 @property(nonatomic, nullable, readonly) UIColor *accent700;
 
+/**
+ Returns the next darker color within the Palette and section (tint or accent). If no darker color
+ exists or @c color is not a predefined palette color, then @c nil is returned.
+
+ @param color used as the base color to perform a palette search for the next darker color.
+
+ @returns the next darker color in the palette, if it exists.
+ */
++ (nullable UIColor *)nextDarkerColorInPaletteForColor:(nullable UIColor *)color;
```

Comment:
Why are these class methods? The implementation that searches all possible palettes seems problematic and doesn't work with user-defined palettes. Recommend making these normal methods.

Reply:
If you absolutely have to use this kind of logic in a situation where you don't already know the right palette to search, then I would recommend breaking out a method that can search a set of palettes for a closest match to a target color. Then the documentation for that method can warn about the performance impact, what happens if there are multiple closest matches, etc.

Reply:
I agree that it only works with predefined methods and the search is undesirable. If we place these methods on a specific palette, how will mutators or views know which palette a color came from?  The original proposal was a category on UIColor, (which could have worked better with custom palettes) but attaching Palette-related data to UIColors was considered undesirable.

Do you think it's better for clients to send palette objects to each view, or did you have another approach in mind?

Reply:
For the performance, I wrote a simple test that iterates over several colors and attempts a lookup for lighter/darker colors of each.

```objectivec
for (int r = 0; r <= 255; r += 8) {
    for (int g = 0; g <= 255; g += 8) {
      for (int b = 0; b <= 255; b += 8) {
        @autoreleasepool {
          UIColor *color = [UIColor colorWithRed:(CGFloat)(r/255.0)
                                           green:(CGFloat)(g/255.0)
                                            blue:(CGFloat)(b/255.0)
                                           alpha:1];
          UIColor *darker = [MDCPalette nextDarkerColorInPaletteForColor:color];
          UIColor *lighter = [MDCPalette nextLighterColorInPaletteForColor:color];
          if (darker || lighter) {
            ++paletteColorsFound;
          }
          ++totalColorsFound;
        }
      }
    }
  }
```

2018-03-14 15:36:02.900840-0400 MDCCatalog[2874:7883980] Palette colors found 1/32768: 34.688088
2018-03-14 15:36:03.482343-0400 MDCCatalog[2874:7883980] Palette colors found 1/32768: 0.581365

This aligns with Time Profiler results on the same device (32.8 seconds):

<img width="1025" alt="timeprofiler-iphone5c-32k_colors" src="https://user-images.githubusercontent.com/1753199/37427256-3ce9ab14-279f-11e8-9a3c-72740d2b90e5.png">

The average is 1 ms to determine the darker/lighter colors of an arbitrary color the first time. Subsequent lookups are cached (until the cache is invalidated) and are around 17 µs.

Reply:
If a color came from a palette, then making code guess the palette after the fact seems a bit silly. If you need a palette in a method, pass the palette into that method.

I would like the following changes to this code:
1. Switch the `nextDarkerColorInPaletteForColor:` and friends to work on an instance of MDCPalette. The user is required to know the palette and we won't be guessing it.
2. If you think guessing a closest known palette from an input color is still valuable, then create a completely separate implementation of that can take an input color and a set of palettes and return an appropriate palette that can be used with `nextDarkerColorInPaletteForColor:`, etc.

Reply:
I can make the change to (1), thanks for the feedback.

Regarding (2), I think it's still valuable. Consider a team that builds VCs that are focused on layout, presentation, binding data to the view, etc.  In those cases they make a simple call to an accessibility mutator (or something similar) that modifies the colors in a view.  Instead of the Palette(s) needing to be injected all the way along the view hierarchy, a quick search can be performed to identify a matching color and then mutate using that if needed.

Reply:
Yep, ok. So if you think 2) is valuable, I would expect an interface like this on MDCPalette:

`+ (??)paletteAndClosestColorFromPalettes:(NSArray<MDCPalette>*)palettes andInputColor:(UIColor *)inputColor`

This would return a pair of `<MDCPalette, MDCPaletteTint>` that indicate the closest palette entry found in any of  `palettes`. (Returning this pair is a bit clunky, but the information coming out is the same.)

The key is to not bake the known set of palettes into this method, because that's just a giant global variable.

You can also provide a _convenience_ method that does search the set of known palettes, but that's clearly documented as a convenience and not the main method.